code/game/objects/items/weapons/trays.dm

```diff
@@ -1,205 +1,159 @@
-	if(istype(A, /obj/item/storage/)) // There used to be here where it would just deny the tray storage if it had contents. It seems wiser, considering just how useful this tray is as a weapon, to deny it backpacks entirely without actually raising its weight class.
-		to_chat(user, "<span class='warning'>The tray won't fit in [A].</span>")
-		return
-	else
-		. = ..()
 
-/obj/item/tray/attack(mob/living/carbon/M, mob/living/carbon/user)
-	user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+// Use the tray in-hand to dump out all its items.
+/obj/item/tray/attack_self(mob/living/user)
+	if (LAZYLEN(carrying))
+		var/turf/T = get_turf(user)
+		overlays.Cut()
+		for (var/obj/item/carried in carrying)
+			carried.dropInto(T)
+			LAZYREMOVE(carrying, carried)
+		user.visible_message(SPAN_NOTICE("[user] dumps out \the [src]."), SPAN_NOTICE("You empty out \the [src]."))
+		return TRUE
+	. = ..()
+
+// When hitting people with the tray, drop all its items everywhere. You jerk.
+/obj/item/tray/attack(mob/living/M, mob/living/user)
+	if (user.a_intent != I_HURT)
+		return FALSE
+	. = ..()
 	// Drop all the things. All of them.
 	overlays.Cut()
 	for(var/obj/item/I in carrying)
-		I.dropInto(M.loc)
+		I.dropInto(get_turf(M))
 		carrying.Remove(I)
-		if(isturf(I.loc))
-			spawn()
-				for(var/i = 1, i <= rand(1,2), i++)
-					if(I)
-						step(I, pick(NORTH,SOUTH,EAST,WEST))
-						sleep(rand(2,4))
-
-
-	if((MUTATION_CLUMSY in user.mutations) && prob(50))              //What if he's a clown?
-		to_chat(M, "<span class='warning'>You accidentally slam yourself with the [src]!</span>")
-		M.Weaken(1)
-		user.take_organ_damage(2)
-		if(prob(50))
-			playsound(M, 'sound/items/trayhit1.ogg', 50, 1)
-			return
-		else
-			playsound(M, 'sound/items/trayhit2.ogg', 50, 1) //sound playin'
-			return //it always returns, but I feel like adding an extra return just for safety's sakes. EDIT; Oh well I won't :3
+		step(I, pick(NORTH, SOUTH, EAST, WEST, NORTHWEST, NORTHEAST, SOUTHWEST, SOUTHEAST))
 
-	var/mob/living/carbon/human/H = M      ///////////////////////////////////// /Let's have this ready for later.
 
+// Bash a rolling pin against a tray like a true knight!
+/obj/item/tray/attackby(obj/item/W, mob/living/user)
+	if(istype(W, /obj/item/material/kitchen/rollingpin) && user.a_intent == I_HURT)
+		if(bash_cooldown < world.time)
+			user.visible_message("<span class='warning'>[user] bashes [src] with [W]!</span>")
+			playsound(user.loc, 'sound/effects/shieldbash.ogg', 50, 1)
+			bash_cooldown = world.time + 25
+		return TRUE
+	else if (user.a_intent != I_HURT && !istype(W, /obj/item/projectile) && !istype(W, /obj/item/clothing))
+		if (calc_carry() + storage_cost_for_item(W) > max_carry)
+			to_chat(user, SPAN_WARNING("\The [src] can't fit \the [W]!"))
+		else if (!can_add_item(W))
+			to_chat(user, SPAN_WARNING("\The [src] can't hold \the [W]!"))
+		else
+			to_chat(user, SPAN_NOTICE("You add \the [W] to \the [src]."))
+			user.drop_item()
```

Comment:
Do not use this proc. Use the unequip variant, and check that it worked before giving the message.